%!TEX root = ../thesis.tex
%*******************************************************************************
%****************************** Third Chapter **********************************
%*******************************************************************************
\chapter{Protocols and analysis of single-cell DNA methylation data}

% **************************** Define Graphics Path **************************
\ifpdf
    \graphicspath{{Chapter3/Figs/Raster/}{Chapter3/Figs/PDF/}{Chapter3/Figs/}}
\else
    \graphicspath{{Chapter3/Figs/Vector/}{Chapter3/Figs/}}
\fi

Our undestanding of DNA methylation has been revolutionized by the development of BS-seq, which offers single-cytosine resolution and absolute quantification of methyl-cytosine genome-wide. Recent advances have demonstrated the power of single-cell sequencing to deconvolve mixed cell populations~\citep{jain_supervised_2007,deng_single-cell_2014,macaulay_single_2014}. Incorporating epigenetic information into this single-cell arsenal will provide insights into epigenetic heterogeneity and transform our understanding of gene regulation.

The first section of this chapter describes scBS-seq, a protocol for genome-wide profiling of DNA methylation in single cells, a statistical method to assess methylation heterogeneity between cells, and applications in embryonic stem cells. The work is based on \citet{smallwood_single-cell_2014}, which was joint work of Sebastien Smallwood, Heather Lee, Christof Angermueller, Felix Krueger, Heba Saadeh, Julian Peat, Simon Andrews, Oliver Stegle, and Wolf Reik.

\begin{center}
\begin{minipage}{.9\linewidth}
\underline{Individual contributions}: Sebastien Smallwood and Heather Lee designed the study, prepared scBS-seq libraries, analyzed data and wrote the manuscript. Felix Krueger, Heba Saadeh, and Sebastien Smallwood performed sequence mapping and analyzed data. Julian Peat contributed to technical developments. Christof Angermueller and Oliver Stegle analysed the data.
\end{minipage}
\end{center}

The second section describes the scM\&T-seq protocol for parallel profiling of DNA methylation and gene expression in single cells, methods to quantify associations between DNA methylation and gene expression, and applications to mouse embryonic stem cells. The work is based on \citet{angermueller_parallel_2016}, which was joint work of Christof Angermueller, Stephen Clark, Heather Lee, Iain Macaulay, Mabel Teng, Tim Xiaoming Hu, Felix Krueger, Sebastien Smallwood, Chris Ponting, Thierry Voet, Gavin Kelsey, Oliver Stegle, and Wolf Reik.

\begin{center}
\begin{minipage}{.9\linewidth}
\underline{Individual contributions}: Christof Angermueller performed all statistical analyses of the data. Heather Lee, Iain Macaulay, Stephen Clark, and Sebastien Smallwood developed the protocol and performed experiments. Heather Lee, Iain Macaulay, Christof Angermueller, Stephen Clark, Oliver Stegle, Wolf Reik, and Chris Ponting interpreted the results. Mabel Teng contributed to method development. Tim Xiaoming Hu processed RNA-seq data. Felix Krueger processed BS-seq data. Wolf Reik, Gavin Kelsey, Iain Macaula, and Thierry Voet contributed protocols and reagents. Heather Lee, Iain Macaulay, Wolf Reik, and Thierry Voet conceived the project.
\end{minipage}
\end{center}
